feat: register db.sql metrics

deniseli opened the pull request:
Fixes https://github.com/TBD54566975/ftl/issues/2285

Register the standard otel database metrics using: https://github.com/XSAM/otelsql
Found call sites by grepping for `sql.Open` and skipping all test code.

Sample:
```
ScopeMetrics #2
ScopeMetrics SchemaURL: 
InstrumentationScope github.com/XSAM/otelsql 0.32.0

Metric #0
Descriptor:
     -> Name: db.sql.latency
     -> Description: The latency of calls in milliseconds
     -> Unit: ms
     -> DataType: Histogram
     -> AggregationTemporality: Cumulative

HistogramDataPoints #0
Data point attributes:
     -> method: Str(sql.conn.begin_tx)
     -> status: Str(ok)
StartTimestamp: 2024-08-08 23:18:01.011762 +0000 UTC
Timestamp: 2024-08-08 23:18:11.012177 +0000 UTC
Count: 23
Sum: 17.082795
Min: 0.137584
Max: 3.233917
```

alecthomas commented on cmd/ftl-controller/main.go:
```diff
@@ -51,7 +52,9 @@ func main() {
 
 	// The FTL controller currently only supports DB as a configuration provider/resolver.
-	conn, err := sql.Open("pgx", cli.ControllerConfig.DSN)
+	conn, err := otelsql.Open("pgx", cli.ControllerConfig.DSN)
```
That is a nice clean API.

alecthomas commented on cmd/ftl/cmd_box_run.go:
```diff
@@ -2,13 +2,14 @@ package main
+	"github.com/XSAM/otelsql"
 	_ "github.com/jackc/pgx/v5/stdlib" // pgx driver
 	"github.com/jpillora/backoff"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
```
We have a test  in internal/observability/client_test.go to ensure the semconv version is consistent that you'll need to update to mention this file, _or_ alternatively change the wording to be more generalised.

alecthomas commented on cmd/ftl/cmd_serve.go:
```diff
@@ -15,8 +14,10 @@ import (
+	"github.com/XSAM/otelsql"
 	"github.com/alecthomas/types/optional"
 	_ "github.com/jackc/pgx/v5/stdlib" // pgx driver
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
```
We have a test  in internal/observability/client_test.go to ensure the semconv version is consistent that you'll need to update to mention this file, _or_ alternatively change the wording to be more generalised.

Probably now there are N locations we should just make the wording generalised.

deniseli replied:
done!

alecthomas commented on internal/modulecontext/database.go:
```diff
@@ -5,6 +5,9 @@ import (
 	"strconv"
 
+	"github.com/XSAM/otelsql"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
```
We have a test  in internal/observability/client_test.go to ensure the semconv version is consistent that you'll need to update to mention this file, _or_ alternatively change the wording to be more generalised.

(same)

alecthomas reviewed (approved):
LGTM with some comments that need to be addressed.

deniseli commented:
Looks like this is breaking tests right now because modulecontext's `Build()` calls internal/reflect's `DeepCopy`, which doesn't expect any values of kind=string that can't be cast to a string. The databases now include a bunch of `metric.descOpt{...}`s wrapping the string descriptions of each metric, so that's confusing the cloning. I'll pick this back up in the morning

deniseli commented:
The alias in question: https://github.com/open-telemetry/opentelemetry-go/blob/c609b12d9815bbad0810d67ee0bfcba0591138ce/metric/instrument.go#L39

deniseli commented:
New changes: 2 minor additions to internal/reflect/reflect.go:

* Added support for string aliases. Before, this would cast the `src` string as long as its `Kind` was `string`, but that would cause a panic when the true type was an alias. In this case, it was [`metric.descOpt`](https://github.com/open-telemetry/opentelemetry-go/blob/c609b12d9815bbad0810d67ee0bfcba0591138ce/metric/instrument.go#L39).
* Added support for the [`list.List`](https://pkg.go.dev/container/list#List) type, which is used by the `global.meter` type inside `otelsql.otConnector`. Before, `DeepCopy` would panic whenever a `list.List` was given as input with `reflect.Set: value of type *list.List is not assignable to type *list.Element`. This adds a special case for that type and performs a standard linked-list deep copy, which is easier than managing consistency across all the backpointers and parent pointers in that structure.

alecthomas commented on ftl-project.toml:
```diff
@@ -1,8 +1,6 @@
 name = "ftl"
 module-dirs = ["examples/go"]
 ftl-min-version = ""
-hermit = false
```
Why'd these get removed?

deniseli replied:
oh what the heck! I'll undo all the changes to that file. Thanks for catching that
